source/agora/utils/TxBuilder.d

```diff
@@ -272,29 +272,35 @@ public struct TxBuilder
         this.data.lock_height = lock_height;
 
         // Finalize the transaction by adding inputs
+        Input[] inputs;
         foreach (ref in_; this.inputs)
-            this.data.inputs ~= Input(in_.hash, Unlock.init, unlock_age);
+            inputs ~= Input(in_.hash, Unlock.init, unlock_age);
 
-        foreach (ref o_; this.data.outputs)
-            o_.type = outputs_type;
+        Output[] outputs;
+        foreach (ref o; this.data.outputs)
+            outputs ~= Output(outputs_type, o.lock, o.value);
 
         // Add the refund output if needed
         if (this.leftover.value > Amount(0))
         {
             if (outputs_type == OutputType.Freeze && this.data.outputs.length == 0) // Single freeze output must be frozen
-                this.data.outputs = [ Output(this.leftover.value, this.leftover.lock, OutputType.Freeze) ];
+                outputs = [ Output(this.leftover.value, this.leftover.lock, OutputType.Freeze) ];
             else
-                this.data.outputs = [ Output(this.leftover.value, this.leftover.lock, OutputType.Payment) ] ~ this.data.outputs;
-            this.data.outputs.sort;
+            {
+                outputs = [ Output(this.leftover.value, this.leftover.lock, OutputType.Payment) ] ~ outputs;
+                outputs.sort;
+            }
         }
         this.data.payload = data;
 
         // Get the hash to sign
         const txHash = this.data.hashFull();
+        inputs.sort;
```

Comment:
```suggestion
        // Get the hash to sign
        inputs.sort;
        outputs.sort;
        const txHash = this.data.hashFull();
```

Shouldn't those 2 lines be enough to fix Faucet ??

Reply:
Also, this is generating an invalid signature, or am I missing something ?

Reply:
The `txHash` here is actually not used as the `hash`ing is done by the unlocker delegate. I will remove the dead variable.